src/lib.ts

```diff
@@ -683,3 +687,80 @@ export function printEventLog(): void {
     );
   }
 }
+
+const rateCache = new Map<string, Map<Monster, number>>();
+
+export function barfEncounterRate(options: {
+  snapper?: boolean;
+  snapperPhylum?: Phylum;
+  olfact?: Monster;
+  longCon?: Monster;
+  motif?: Monster;
+  turtle?: Monster;
+  monkeyPoint?: Monster;
+  humanity?: boolean;
+}): Map<Monster, number> {
+  const olfact = options.olfact ?? get("olfactedMonster");
+  const longCon = options.longCon ?? get("longConMonster");
+  const motif = options.motif ?? get("motifMonster");
+  const turtle = options.turtle ?? get("_gallapagosMonster");
+  const monkeyPoint = options.monkeyPoint ?? get("monkeyPointMonster");
+  const snapper = options.snapper ?? myFamiliar() === $familiar`Red-Nosed Snapper`;
+  const snapperPhylum = options.snapperPhylum ?? Snapper.getTrackedPhylum();
+  const humanity = options.humanity ?? have($effect`Ew, The Humanity`);
+
+  const zoneMonsters = $monsters`garbage tourist, angry tourist, horrible tourist family`;
+  const encounterQueue = zoneMonsters.filter((m) =>
+    $location`Barf Mountain`.combatQueue.includes(`${m}`),
+  );
+
+  const cacheKey = [
+    olfact,
+    longCon,
+    motif,
+    turtle,
+    monkeyPoint,
+    snapper,
+    snapperPhylum,
+    humanity,
+    ...encounterQueue,
+  ]
+    .map((v) => `${v}`)
+    .join(":");
+
+  const cachedValue = rateCache.get(cacheKey);
+  if (cachedValue) {
+    return cachedValue;
+  }
+
+  const copies = (target: Monster | null, n: number): Monster[] =>
+    n === 0 ? [] : [...zoneMonsters.filter((m) => m === target), ...copies(target, n - 1)];
+
+  const monsterQueue = [
+    ...zoneMonsters,
+    ...copies(olfact, 3),
+    ...copies(longCon, 3),
+    ...copies(motif, 2),
+    ...copies(turtle, 1),
+    ...copies(monkeyPoint, 2),
+    ...zoneMonsters
+      .filter((m) => snapper && m.phylum === snapperPhylum)
+      .flatMap((m) => copies(m, 2)),
+    ...zoneMonsters
+      .filter((m) => humanity && m.phylum === $phylum`dude`)
+      .flatMap((m) => copies(m, 2)),
+  ];
+
+  const encounters = monsterQueue.flatMap((m) =>
```

Comment:
I don't understand what this variable is meant to represent

Reply:
update: I think I might understand it, but if I do then it's not quite right: we only reroll 75% of the time, but even a reroll can reroll.

This math is really hard to do--slaw wrote a big ol' thesis about it. I think we have to do it numerically instead of analytically.